### DISCUSSION TITLE
Modbus Bridge - should there always be a response?

### DISCUSSION THREAD

**Discussion by jpadie** [arendst/Tasmota#19137] [Q&A]
I'm experimenting with the modbus RTU bridge with a known good chipset and a known good modbus slave.  

I'm wondering if we should always expect a result from the modbussend command?  even if it is a time out or an error.  Currently I get nothing beyond the "Done".  Using an ESP32 wroom with tasmota 13.0.0 web-installed.

```
20:19:40.400 -> 19:19:39.987 CMD: modbusserialconfig 3
20:19:40.400 -> 19:19:39.988 SRC: Serial
20:19:40.400 -> 19:19:39.990 CMD: Grp 0, Cmd 'MODBUSSERIALCONFIG', Idx 1, Len 1, Pld 3, Data '3'
20:19:40.439 -> 19:19:40.002 MQT: rab/poolManager/stat/RESULT = {"ModbusSerialConfig":"8N1"}
20:19:43.866 -> 19:19:43.453 WIF: Checking connection...
20:19:45.390 -> 19:19:44.969 CMD: modbusbaudrate 19200
20:19:45.390 -> 19:19:44.970 SRC: Serial
20:19:45.390 -> 19:19:44.972 CMD: Grp 0, Cmd 'MODBUSBAUDRATE', Idx 1, Len 5, Pld 19200, Data '19200'
20:19:45.390 -> 19:19:44.983 MQT: rab/poolManager/stat/RESULT = {"ModbusBaudrate":19200}
20:19:50.554 -> 19:19:50.137 CMD: modbussend {"deviceAddress":1, "functionCode":3, "startAddress":8200, "type":"hex","count":1}
20:19:50.554 -> 19:19:50.139 SRC: Serial
20:19:50.554 -> 19:19:50.140 CMD: Grp 0, Cmd 'MODBUSSEND', Idx 1, Len 82, Pld -99, Data '{"deviceAddress":1, "functionCode":3, "startAddress":8200, "type":"hex","count":1}'
20:19:50.612 -> 19:19:50.179 MQT: rab/poolManager/stat/RESULT = {"ModbusSend":"Done"}
20:20:03.869 -> 19:20:03.446 WIF: Checking connection...
20:20:12.753 -> 19:20:12.345 CMD: modbussend {"deviceAddress":1, "functionCode":4, "startAddress":8200, "type":"hex","count":1}
20:20:12.753 -> 19:20:12.347 SRC: Serial
20:20:12.805 -> 19:20:12.349 CMD: Grp 0, Cmd 'MODBUSSEND', Idx 1, Len 82, Pld -99, Data '{"deviceAddress":1, "functionCode":4, "startAddress":8200, "type":"hex","count":1}'
20:20:12.805 -> 19:20:12.387 MQT: rab/poolManager/stat/RESULT = {"ModbusSend":"Done"}
20:20:21.454 -> 19:20:21.043 CMD: modbussend {"deviceAddress":1, "functionCode":4, "startAddress":8200, "type":"raw","count":1}
20:20:21.454 -> 19:20:21.045 SRC: Serial
20:20:21.454 -> 19:20:21.047 CMD: Grp 0, Cmd 'MODBUSSEND', Idx 1, Len 82, Pld -99, Data '{"deviceAddress":1, "functionCode":4, "startAddress":8200, "type":"raw","count":1}'
20:20:21.503 -> 19:20:21.086 MQT: rab/poolManager/stat/RESULT = {"ModbusSend":"Done"}

```

**Comment by pkkrusty** [comment]
If you're trying to track down Modbus issues, using Smart Meter Interface with scripting is helpful. The dump mode lets you see what is being sent and what Tasmota is seeing in response. Is your RS485 module blinking both send and receive lights?

**Reply by jpadie** [reply]
there is no module - it's a max485 chip on the same board as the esp32 wroom module, directly connected to pins 17 (RE, RE_), 2 (DI) and 15 (RO).  

as it happens I am getting a secondary issue in that pin assignment seems to prevent the esp32 from connecting to any AP.  remove the assignments and all is well.  

can you elaborate re using smart meter interface?  at the moment I'm connected to an oscilloscope to see whether there is any traffic at all.  

(also what I'm really looking for is a view on whether there should always be some response back from tasmota beyond the "done".  a result or an error etc)

**Reply by pkkrusty** [reply]
Normally there should be a result message printed out, whether it's an error or good data. 

What config pin assignments are you using? ModBr Tx, ModBr Rx and ModBr Tx Ena? A recent discussion indicated some issues with the Enable pin not doing what it's supposed to. https://github.com/arendst/Tasmota/discussions/15703#discussioncomment-6192150

https://tasmota.github.io/docs/Smart-Meter-Interface/

Not sure how things work with Enable pin.  Most Modbus stuff for Tasmota assumes a RS485 module with auto flow control. See https://github.com/arendst/Tasmota/discussions/11504#discussioncomment-588391

**Reply by jpadie** [reply]
cheers.

interesting that there is no result nor error then.

checking the codebase for 13.0.0.2, it looks like the enable pin is being asserted if the definition of TASMOTA_MODBUS_TX_ENABLE is present, which is the case by default.   It is asserted LOW on starting comms; HIGH just before the packet is sent, then LOW again.  this looks correct and I don't think there should be a race condition (maybe too much capacitance on the pin?).   I've tried with and without pullups (10k) to 3.3v and pulldowns (10k) to Gnd. 

for pin assignments I'm using GPIO 2 for  ModBr Tx (DI), 15 for ModBr Rx (RO) and 17 for ModBr Tx Ena (RE, tied to !RE).

the bit that is worrying me most is that there is no feedback from the tasmota driver even for a time out.

**Comment by Jason2866** [comment]
idk if the chip max485 is working together with Tasmota. The supported and verified working board is this one
https://github.com/jarcikw/SDM120_Homeassistant/wiki/SDM120-&-nodemcu-&-RS485-wiring

There is feedback for other converter chips NOT working.

EDIT1:
> there is no module - it's a max485 chip on the same board as the esp32 wroom module, directly connected to pins 17 (RE, RE_), 2 (DI) and 15 (RO).

GPIO 17 can make issues. The GPIO is used to connect PSRAM. Tasmota enables PSRAM and does check for PSRAM. If this fails the GPIO is reseted for normal use. In the test phase the GPIO 17 changes often it's state.
This can irritate chips connected to this GPIO. Maybe a init of the chip after this is needed.
Long story short. Avoiding GPIO17 is a good strategy since this one of the default GPIOs for connecting PSRAM.

So when starting with Tasmota and RS485 it is highly recommended to start with the setup described in the documentation. After this works as wanted, hardware changes can be tried.

EDIT2: The max485 chip is verified NOT working with Tasmota.

**Reply by jpadie** [reply]
Thank you Jason 

I'm not sure why the max485 chip would not work where the full duplex variant would.   There is chat in the GitHub going back a number of years about the max485 chip being used in various solutions.   Perhaps not directly with the modbus bridge module though.  

Is there a canonical statement of why the max485 chip is not functional with tasmota?

**Reply by Jason2866** [reply]
Yes, search in discussion with rs485

**Reply by jpadie** [reply]
have read a number of the more recent issues (2019) and discussions.  Nowhere can I see that the max485 chip canonically does not work with tasmota.  The inferences are actually the opposite (backed up as well by [Theo's messages](https://github.com/arendst/Tasmota/pull/17247) in Dec 2022.  it's the max485 (and other half-duplex clones) that need the tx Enable.  

there is the separate issue of the max485 chip being 5v but that's entirely separate.

**Reply by Jason2866** [reply]
https://github.com/arendst/Tasmota/discussions/10795#discussioncomment-330290

**Comment by Jason2866** [comment]
Ahh, overlooked https://github.com/arendst/Tasmota/commit/bbde894628832583a52648ebbe3f95bd86438b09
Max485 should work with this change.

My recommendation still is to start with the known working setup.

**Reply by jpadie** [reply]
I did try earlier today.  No joy.  But I also didn't see any led activity on the external module so potentially fried.

I'm not a believer in the gpio2 issue as source until someone shows me the architecture that in conjunction with tasmota creates the scenario.  I use gpio2 too often in other scenarios without hassle (with and without tasmota).  I will test it but I am now on my way home to the UK and won't be back in France (where the test rig is) until late August.

**Reply by pkkrusty** [reply]
I had an instance where my Rx pin was fried, and I was confused why I wasn't receiving a response, even though Tx/Rx lights were blinking on the module. Once I swapped it, everything worked as expected. From your oscilloscope, could you tell if your lines were sending and receiving and flow controlling correctly? I have a $8 eBay logic analyzer that has been immensely helpful in troubleshooting that kind of thing.
